### ISSUE TITLE
Missing access mode: discard_write

### ISSUE THREAD

**Issue by j-stephan** [triSYCL/sycl#31]
Using the same N-body code as in #28 I receive the following failed assertion once the control flow reaches the first host-to-device copy:

```
nbody.sw_emu: /home/jstephan/software/sycl/lib/clang/9.0.0/include/CL/sycl/detail/buffer_impl.hpp:519: size_t cl::sycl::detail::buffer_impl<std::allocator<char> >::convertSycl2OCLMode(cl::sycl::access::mode) [AllocatorT = std::allocator<char>]: Assertion `0 && "Unhandled conversion from Sycl access mode to OCL one."' failed.
```

And the mode is indeed missing from `buffer_impl.hpp`. Is this a feature waiting for implementation? I haven't found anything related in the issues here or the original triSYCL repository.

Edit: Ah, I should've read @agozillon's message more thoroughly. I suppose this is the "runtime assertion" failing. This already seems to be fixed in Intel's implementation:

```
template <typename AllocatorT>
size_t
buffer_impl<AllocatorT>::convertSycl2OCLMode(cl::sycl::access::mode mode) {
  switch (mode) {
  case cl::sycl::access::mode::read:
    return CL_MEM_READ_ONLY;
  case cl::sycl::access::mode::write:
  case cl::sycl::access::mode::discard_write:
    return CL_MEM_WRITE_ONLY;
  case cl::sycl::access::mode::read_write:
  case cl::sycl::access::mode::discard_read_write:
  case cl::sycl::access::mode::atomic:
    return CL_MEM_READ_WRITE;
  default:
    assert(0 && "Unhandled conversion from Sycl access mode to OCL one.");
    return 0;
  }
}
```
```

**Comment by agozillon** [comment]
Once this is merged into Unified/Next https://github.com/triSYCL/sycl/pull/30 (should be end of Today or tomorrow PDT), I believe discard_write should exist and Unified/Next should be somewhat 'stable', I'll double check though and add an edit note if i am incorrect. The master branch is a fair bit behind the main Intel SYCL branch at the moment unfortunately.

**Comment by agozillon** [comment]
Going to close this as this should now exist in the unified-next branch (if it doesn't or you have some other problems with it, feel free to reopen this issue or create a new one).

However, hw_emu is still a little ropy on that branch unfortunately at least with the new SDAccel 2019.1 release, we're working on it, but if you encounter any new bugs please do open an issue.

Edit: Should also mention that it should also have sycl vec support now, I've not had a chance to test them out too thoroughly though but I'm aware that some of the vec math operations have similar missing symbol issues to the basic math functions.

**Comment by j-stephan** [comment]
Okay, I will upgrade to `sycl/unified/next` and 2019.1. Is the next-branch the development branch or will it be merged into `sycl/unified/master` in the future?

**Comment by agozillon** [comment]
It's the development branch for just now, but it will eventually be merged into **sycl/unified/master** or replace it, at least that's my intent for just now.
